Modules/_asynciomodule.c

```diff
@@ -2049,13 +2067,15 @@ unregister_task(asyncio_state *state, TaskObj *task)
         assert(state->asyncio_tasks.head != task);
         return;
     }
+    ASYNCIO_STATE_LOCK(state);
```

Comment:
I didn't realize it previously, but I think the `if (task->next == NULL) {` needs to be within the lock. Otherwise, there's a potential data race on `task->next` if another thread is concurrently removing either the task or the task next to it in the list.

I think to make the code more clear, we should just put the whole function within the lock, but be careful about the `return` like with `register_task`.